service-base/src/test/java/org/eclipse/hono/service/tenant/BaseTenantServiceTest.java

```diff
@@ -0,0 +1,170 @@
+    }
+
+    /**
+     * Verifies that the base service accepts a request for adding
+     * a tenant that contains the minimum required properties.
+     */
+    @Test
+    public void testAddSucceedsForMinimalData() {
+
+        final JsonObject testPayload = createValidTenantPayload();
+
+        final Message<JsonObject> msg = createMessageMockForPayload(Action.ACTION_ADD, testPayload);
+        tenantService.processTenantMessage(msg);
+
+        verify(msg).reply(resultWithStatusCode(HTTP_CREATED, TEST_TENANT));
+    }
+
+    /**
+     * Verifies that the base service accepts a request for adding
+     * a tenant that contains the minimum required properties.
+     */
+    @Test
+    public void testAddFailsForIncompleteMessage() {
+
+        final JsonObject testPayload = createValidTenantPayload();
+
+        final Message<JsonObject> msg = createIncompleteMessageMockForPayload(testPayload);
+        tenantService.processTenantMessage(msg);
+
+        verify(msg).reply(resultWithStatusCode(HTTP_BAD_REQUEST, null));
+    }
+
+    /**
+     * Verifies that the base service fails for a payload that defines an empty adapter array (must be null or has to
+     * contain at least one element).
+     */
+    @Test
+    public void testAddFailsForEmptyAdapterArray() {
+
+        final JsonObject testPayload = createValidTenantPayload();
+        testPayload.put(TenantConstants.FIELD_ADAPTERS, new JsonArray());
+
+        final Message<JsonObject> msg = createMessageMockForPayload(Action.ACTION_ADD, testPayload);
+        tenantService.processTenantMessage(msg);
+
+        verify(msg).reply(resultWithStatusCode(HTTP_BAD_REQUEST, TEST_TENANT));
+    }
+
+    /**
+     * Verifies that the base service fails for a payload that defines an adapter entry, but does not provide the
+     * mandatory field {@link TenantConstants#FIELD_ADAPTERS_TYPE}.
+     */
+    @Test
+    public void testAddFailsForAdapterConfigWithoutType() {
+
+        final JsonObject testPayload = createValidTenantPayload();
+        final JsonArray adapterArray = new JsonArray();
+        // no type specified (which is a violation of the API)
+        adapterArray.add(new JsonObject());
+        testPayload.put(TenantConstants.FIELD_ADAPTERS, adapterArray);
+
+        final Message<JsonObject> msg = createMessageMockForPayload(Action.ACTION_ADD, testPayload);
+        tenantService.processTenantMessage(msg);
+
+        verify(msg).reply(resultWithStatusCode(HTTP_BAD_REQUEST, TEST_TENANT));
+    }
+
+    @SuppressWarnings("unchecked")
+    private static Message<JsonObject> createMessageMockForPayload(final Action operation, final JsonObject payload) {
+
+        JsonObject requestBody = new JsonObject();
+        requestBody.put(RequestResponseApiConstants.FIELD_TENANT_ID, TEST_TENANT);
+        requestBody.put(MessageHelper.SYS_PROPERTY_SUBJECT, operation);
+        if (payload != null) {
+            requestBody.put(CredentialsConstants.FIELD_PAYLOAD, payload);
```

Comment:
IMHO we should not refer to `CredentialsConstants` here

Reply:
Definitely, thanks!